### PR TITLE
Fast UniswapPoolRow

### PR DESCRIPTION
**Pull request by tchayen**
Fixes TEAM1-36

Based on Michał's PR #3367. Unfortunately due to conflicts and some differences in implementation between changes made by Terry and Michał roughly around the same time, I decided it's much easier to just manually pick relevant changes and start a new branch (I tried rebasing but I gave up after 30 mins of picking between sets of changes irrelevant to the PR).

## What changed (plus any additional context for devs)

Uniswap pool row got a new implementation.

## PoW (screenshots / screen recordings)

See comments for most up to date screenshots.

## Dev checklist for QA: what to test

Check that uniswap pools work the same way as they did.

## Final checklist

- [x] Assigned individual reviewers?
- [x] Added labels?
- [x] Added e2e tests? if not please specify why **Reimplementation of existing functionality which is either covered by tests or out of scope of this PR**
- [x] If you added new files, did you update the CODEOWNERS file?

### CODE REVIEW

**Comment by tchayen** [comment]
Oh yea, there was quite a bit of a difference. The only remaining difference is in text rendering but I think this is in big part related to the text component in design system and there are limited options to make it match exactly 1:1 (but I think this is fine and partially expected outcome).

| Before | After | Difference |
| -- | -- | -- |
| before | after | diff |

One thing I noticed is the lack of shadows. I need to have a look at it.

**Comment by terrysahaidak** [comment]
I don't agree we should ditch Text from DS, we are already using DS for other Fast Rows with design team approvals.
Also I don't agree on 1:1 match since previous approach to the layout is basically make it look nice on iOS and don't care about Android - this is why it no longer matches the previous design on Android - since finally we render text the same way and look on both iOS and Android. So basically we should make sure we match design on iOS.

**Comment by osdnk** [comment]
The problem is that some things are impossible with a design system even on iOS. In this case, I would hesitate to wait for the designers' approval and split the design performance improvement and improving consistency into two parts. Those are two tasks

File: src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx
Comment (terrysahaidak) [inline src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx:11]: I think we might have somewhere a type for this

File: src/components/asset-list/RecyclerAssetList2/FastComponents/FastPoolValue.tsx
Comment (terrysahaidak) [inline src/components/asset-list/RecyclerAssetList2/FastComponents/FastPoolValue.tsx:62]:
```suggestion
  }, [colors, nativeCurrency, type, value]);

  return (
```

Comment (terrysahaidak) [reply src/components/asset-list/RecyclerAssetList2/FastComponents/FastPoolValue.tsx:62]: Some whitespace to improve code readability

**Review by terrysahaidak — CHANGES_REQUESTED** [review]
Few comments.

Also, could you please wrap `poolData` in `useUniswapPools` with `useDeepMemo` since I remember it was running some pulling and returning same object with new reference every N seconds, that would trigger some background rerenders of the list.

**Comment by terrysahaidak** [comment]
@osdnk consistent text rendering on both iOS and Android is something that Design System's Text component is designed to fix and it does this very well. 

Unfortunately, we can't talk about performance and keep using Text from styled components.

Also, as a reminder, #3313 is already merged and approved by @christianbaroni (also via #3558).

**Comment by osdnk** [comment]
@terrysahaidak I don't think this is a good place for this discussion, but this was actually a problem that 3313 required follow-up with 3558. I express that we can use raw `Text` for `react-native` if we cannot obtain 1:1 consistency with DS. Then migrate to DS in the different PR. We are solving different problems here and we should avoid coupling problems together.

**Comment by tchayen** [comment]
Very good points. I will follow the approach Michał suggested to avoid generating potential issues.

Anyway, changing text to the DS variant afterwards is mostly a quick change so definitely it will be quite fast for us to implement and fast for designers to verify.

**Comment by terrysahaidak** [comment]
I don't agree with the points. We should not introduce yet another concept and add more of these `android ? 2 : 3` into the codebase to fix problems Design System Text component is supposed to fix.

File: src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx
Comment (tchayen) [reply src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx:11]: CC @marco Maybe you know something that can work here?

Comment (marco) [reply src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx:11]: @tchayen Correct, there are a few token types located in [`src/entities/tokens`](https://github.com/rainbow-me/rainbow/blob/aeb8cdfca8eef3ca4b559e46d16fa0b5f2647fb7/src/entities/tokens.ts), however from looking at the code I think the type that actually fits best might be [`UniswapPoolAddressDetailsFull`](https://github.com/rainbow-me/rainbow/blob/b8b4f082473eb7086c312c50045e955853546c48/src/redux/uniswapLiquidity.ts#L35), since the data is ultimately loaded from the `useUniswapLiquidity` hook. Currently that type is not exported, (it is in a future PR but that has not been merged yet), so you might want to export it from that file and then use it?

**Comment by tchayen** [comment]
As discussed in a meeting yesterday, we decided to skip working out a unified way of dealing with the text inconsistencies of pre DS vs DS and instead we will figure it out on a PR basis.

In this case the changes seem to successfully reproduce everything but a minor text rendering differences, so it will be verified and corrected with designers.

File: src/components/discover-sheet/UniswapPoolsSection.js
Comment (tchayen) [reply src/components/discover-sheet/UniswapPoolsSection.js]: Added it but not sure if there are any gotchas about using it. Please verify if I am not doing anything stupid.

**Review by osdnk — APPROVED** [review]
Ok! I approve of this from the code perspective. This works and looks ok.
However, please have the designer's approval.

File: src/hooks/useLatestCallback.ts
Comment (osdnk) [inline src/hooks/useLatestCallback.ts:4]: hey, we rather use `import { useEffect } from 'react`
etc. Not sure why here this is different?

File: src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx
Comment (osdnk) [inline src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx:118]: was this originally? Any chance to obtain this with `margin: -10`? etc?

File: src/hooks/useLatestCallback.ts
Comment (tchayen) [reply src/hooks/useLatestCallback.ts:4]: Following Terry's advice I copied it over from his [PR](https://github.com/rainbow-me/rainbow/pull/3547/files#diff-aa6a3ef706bb0a495227d6c59c5f491546ccf76a5772a4c735238376ab7696aeR92) so I'd rather discuss it there and re-sync here as this is intended to be exactly the same file post-merges.

File: src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx
Comment (tchayen) [reply src/components/asset-list/RecyclerAssetList2/FastComponents/FastUniswapPoolRow.tsx:113]: Ok. Margin works the same. Changing it.

File: src/components/asset-list/RecyclerAssetList2/FastComponents/FastPoolValue.tsx
Comment (tchayen) [reply src/components/asset-list/RecyclerAssetList2/FastComponents/FastPoolValue.tsx:75]: Absolutely! Thanks for catching it!

**Comment by tchayen** [comment]
After playing a bit more with tweaks a bit more today:

| develop | this branch |
|--|--|
|develop|fast|

File: src/components/discover-sheet/UniswapPoolsSection.js
Comment (terrysahaidak) [inline src/components/discover-sheet/UniswapPoolsSection.js:234]: This hook doesn't have dependencies

File: src/components/discover-sheet/UniswapPoolsSection.js
Comment (terrysahaidak) [inline src/components/discover-sheet/UniswapPoolsSection.js:312]: why so?

File: src/components/discover-sheet/UniswapPoolsSection.js
Comment (terrysahaidak) [reply src/components/discover-sheet/UniswapPoolsSection.js:234]: Did you check for ts errors?

**Review by terrysahaidak — APPROVED** [review]
Looks good overall

File: src/components/discover-sheet/UniswapPoolsSection.js
Comment (tchayen) [reply src/components/discover-sheet/UniswapPoolsSection.js:312]: I need to check. Actually this might be unnecessary 😅.

File: src/components/discover-sheet/UniswapPoolsSection.js
Comment (tchayen) [reply src/components/discover-sheet/UniswapPoolsSection.js:234]: I wasn't looking specifically for TS errors and nothing caught my attention. Probably because it's a JS file so IDE didn't complain?

Fixing anyway. Thanks for catching this!

Comment (terrysahaidak) [reply src/components/discover-sheet/UniswapPoolsSection.js]: Oh I just noticed that this is a js file lol, sorry.

**Comment by tchayen** [comment]
Update: on hold, Jeremy is verifying with Christian as there might be small design adjustments.

**Comment by tchayen** [comment]
Updated styles:

|Light|Dark|
|--|--|
|IMG_42F50378B482-1|IMG_B15F7A27C33F-1|

**Comment by BrodyHughes** [comment]
dm'd @tchayen about a couple style things I was unsure about. Functionality seems un-impacted to me as compared to prod.

**Review by BrodyHughes — APPROVED** [review]
QA checked & compared to prod:
- discover 👍 
- balance screen 👍 
- expanded state 👍 
- design 👍

**Pull request merged**
